src/zarr/core/array.py

```diff
@@ -372,9 +372,15 @@ async def open(
         else:
             # V3 arrays are comprised of a zarr.json object
             assert zarr_json_bytes is not None
+            zarr_metadata = json.loads(zarr_json_bytes.to_bytes())
```

Comment:
I'm not too happy with the need for this... Without it, we eventually try to load the metadata as an ArrayMetadata, which throws a `ValueError` IIRC. But the parent does a

```
try:
    open_array(...)
except KeyError:
    open_group(...)
```

Might see if we can clean that up. At least for `v3` we can key off the `node_type` once we've loaded the JSON metadata.